### PR TITLE
Hook constants / Hooks Manager

### PR DESCRIPTION
**Pull request by AdrienClairembault**
These changes make using hooks in plugins easier thanks to some new constants and helper methods.  
Main benefits are auto completion ( = less reliant on documentation) and "easier to look for" references (= allow searching on the const name instead of the string value which may return a lot of unrelated code).  
No impact on existing plugins, the new helper methods brought by `HookManager` are completely optional.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |

### CODE REVIEW

File: inc/plugin/hooks.class.php [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+         self::ADD_JAVASCRIPT,
+      ];
+   }
+
+   /**
+    * Get functionnals hooks
+    *
+    * @return array
+    */
+   public static function getFonctionnalHooks(): array {
```
Comment (cconard96) [inline inc/plugin/hooks.class.php:121]:
```suggestion
   public static function getFunctionalHooks(): array {
```

File: inc/plugin/hookmanager.class.php [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+      $PLUGIN_HOOKS[$hook][$this->plugin][] = $file;
+   }
+
+   /**
+    * Add a functionnal hook
+    *
+    * @param string $hook
+    * @param string $file
+    */
+   public function registerFunctionnalHook(
```
Comment (cconard96) [inline inc/plugin/hookmanager.class.php:98]:
```suggestion
   public function registerFunctionalHook(
```

File: inc/plugin/hookmanager.class.php [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+   public function enableCSRF(): void {
+      $PLUGIN_HOOKS[Hooks::CSRF_COMPLIANT][$this->plugin] = true;
+   }
+
+   /**
+    * Add a given javascript file
+    *
+    * @param string $file
+    */
+   public function registerJavascriptFile(string $file): void {
```
Comment (cconard96) [inline inc/plugin/hookmanager.class.php:55]: Needs to support `add_javascript_module` hooks in addition to the plain `add_javascript` hook.

File: inc/plugin/hookmanager.class.php [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+   public function registerFunctionnalHook(
+      string $hook,
+      callable $function
+   ): void {
+      global $PLUGIN_HOOKS;
+
+      // Check if the given hook is a valid functionnal hook
+      $allowed_file_hooks = Hooks::getFonctionnalHooks();
+      if (!isset($allowed_file_hooks[$hook])) {
+         trigger_error("Invalid functionnal hook: '$hook'", E_USER_ERROR);
```
Comment (cconard96) [inline inc/plugin/hookmanager.class.php:107]:
```suggestion
         trigger_error("Invalid functional hook: '$hook'", E_USER_ERROR);
```

File: inc/plugin/hooks.class.php [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+namespace Glpi\Plugin;
+
+class Hooks
+{
+   // Boolean hooks
+   const CSRF_COMPLIANT = 'csrf_compliant';
+
+   // File hooks
+   const ADD_CSS        = 'add_css';
+   const ADD_JAVASCRIPT = 'add_javascript';
```
Comment (cconard96) [inline inc/plugin/hooks.class.php:42]: Need to add `ADD_JAVASCRIPT_MODULE = 'add_javascript_module'`

**Review by cconard96 — CHANGES_REQUESTED** [review]

**Comment by cconard96** [comment]
Just an FYI, I didn't comment on the usages of the functions with typos, so they will need checked/fixed manually.

File: inc/plugin/hookmanager.class.php (hunk 3 above)
Comment (AdrienClairembault) [reply inc/plugin/hookmanager.class.php:55]: I can't find any references to `add_javascript_module` in the code.
Are you sure it's available on this branch ?

Comment (cconard96) [reply inc/plugin/hookmanager.class.php:55]: My bad. #8970 targeted the modern-ui branch and not master so it isn't in the master branch yet.

**Review by cconard96 — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**